pkg/routing/kvstore/kvstore.go

```diff
@@ -0,0 +1,183 @@
+package kvstore
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/nats.go/jetstream"
+	"github.com/pkg/errors"
+	"github.com/samber/lo"
+
+	"github.com/bacalhau-project/bacalhau/pkg/models"
+	"github.com/bacalhau-project/bacalhau/pkg/routing"
+)
+
+type NodeStoreParams struct {
+	BucketName     string
+	ConnectionInfo interface{}
+}
+
+type NodeStore struct {
+	js jetstream.JetStream
+	kv jetstream.KeyValue
+}
+
+func NewNodeStore(params NodeStoreParams) (*NodeStore, error) {
+	url, ok := params.ConnectionInfo.(string)
+	if !ok {
+		return nil, errors.New("invalid connection info provided to KV Node Store")
+	}
+
+	// The connection we get from NATS is thread-safe (see https://pkg.go.dev/github.com/nats-io/nats.go#Conn)
+	// so no need to wrap everything in a mutex
+	nc, err := nats.Connect(url)
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("failed to connect to nats network at %s", url))
+	}
+
+	js, err := jetstream.New(nc)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to connect to jetstream")
+	}
+
+	bucketName := strings.ToLower(params.BucketName)
+	if bucketName == "" {
+		return nil, errors.New("bucket name is required")
+	}
+	kv, err := js.CreateKeyValue(context.Background(), jetstream.KeyValueConfig{
+		Bucket: bucketName,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create key-value store")
+	}
+
+	return &NodeStore{
+		js: js,
+		kv: kv,
+	}, nil
+}
+
+func (n *NodeStore) FindPeer(ctx context.Context, peerID peer.ID) (peer.AddrInfo, error) {
+	// TODO: Remove this once we now longer need to implement the routing.PeerStore interface
+	// We are temporarily matching the code of the inmemory.NodeStore which never returns an
+	// error for this method.
+	nodeID := peerID.String()
+	info, err := n.Get(ctx, nodeID)
+	if err != nil {
+		return peer.AddrInfo{}, nil
+	}
+
+	return *info.PeerInfo, nil
+}
+
+// Add adds a node info to the repo.
+func (n *NodeStore) Add(ctx context.Context, nodeInfo models.NodeInfo) error {
+	data, err := json.Marshal(nodeInfo)
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal node info adding to node store")
+	}
+
+	_, err = n.kv.Put(ctx, nodeInfo.ID(), data)
+	if err != nil {
+		return errors.Wrap(err, "failed to write node info to node store")
+	}
+
```

Comment:
it seems nats kv supports revisions. Will `Put` append a revision or replace all previous revisions. My concern is we are doing a lot of puts as part of our hearbeats, and might cause issues if we don't cleanup old revisions

Reply:
I'm going to try and keep the heartbeats/resource-usage-updates in memory because they'll be stale as soon as we store them in the KVStore.  The only real time I think we'll need to update is for label changes on the compute node + approval state.